### ISSUE TITLE
No way to change terminal colors with termguicolors set

### ISSUE THREAD

**Issue by balachia** [neovim/neovim#8301]
- `nvim --version`:
```
NVIM v0.2.2
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/local/Homebrew/Library/Homebrew/shims/super/clang -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -I/tmp/neovim-20180209-92407-udkzoo/neovim-0.2.2/build/config -I/tmp/neovim-20180209-92407-udkzoo/neovim-0.2.2/src -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/gettext/include -I/usr/include -I/tmp/neovim-20180209-92407-udkzoo/neovim-0.2.2/build/src/nvim/auto -I/tmp/neovim-20180209-92407-udkzoo/neovim-0.2.2/build/include
Compiled by brew@HighSierra.local

Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/Cellar/neovim/0.2.2_1/share/nvim"

Run :checkhealth for more info
```
- Vim (version: ) behaves differently? `N/A`
- Operating system/version: macOS 10.13.4
- Terminal name/version: kitty
- `$TERM`: xterm-kitty

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC --cmd "set termguicolors"
:terminal
printf '\x1b[31mRED\x1b[0m\n'
<C-\><C-n>:let g:terminal_color_1 = '#654321'
printf '\x1b[31mRED\x1b[0m\n'
```

### Actual and expected behaviour

Terminal colors don't change during terminal lifetime.

### Desired

Per [https://neovim.io/doc/user/nvim_terminal_emulator.html](https://neovim.io/doc/user/nvim_terminal_emulator.html), `g:terminal_color_#' is read only at terminal startup. As expected, attempting to change these variables doesn't cause any color change. Given the discussion at [#4696](https://github.com/neovim/neovim/issues/4696), changes to colors of the underlying terminal emulator also don't propagate into the neovim :terminal.

This makes drastic colorscheme changes impossible. I can use :colorscheme to change from a light background to a dark background neovim but have no way to update terminal colors accordingly, and usually do between day and nighttime. This can make the terminal either extremely unpleasant to use in one of the two cases, or slightly unpleasant to use all of the time.

Is there any way both to use termguicolors in most of neovim while also allowing the terminal to either reflect emulator colors or to have adjustable colors?

**Comment by justinmk** [comment]
can you try https://github.com/neovim/neovim/pull/7406 ?

**Comment by jfelchner** [comment]
@justinmk my bad, sorry for wasting your time here.  I hadn't applied it.  I thought it was already in `master`.

However I _did_ apply the patch and now what I'm seeing is that all terminal output is in "normal" color.

![screenshot 2018-05-16 at 14 38 28](https://user-images.githubusercontent.com/285582/40139862-d5e68c4e-5916-11e8-9523-b525c811658f.png)

It doesn't matter what I set `ctermfg` to, it doesn't apply any colors.  I've tried both named and numbered colors.  This is what I currently have:

```
    " Black and Dark Gray
    hi  TermColor0          ctermfg=2
    hi  TermColor8          ctermfg=2

    " Red
    hi  TermColor1          ctermfg=2
    hi  TermColor9          ctermfg=2

    " Green
    hi  TermColor2          ctermfg=2
    hi  TermColor10         ctermfg=2

    " Yellow
    hi  TermColor3          ctermfg=2
    hi  TermColor11         ctermfg=2

    " Blue
    hi  TermColor4          ctermfg=2
    hi  TermColor12         ctermfg=2

    " Magenta
    hi  TermColor5          ctermfg=2
    hi  TermColor13         ctermfg=2

    " Aqua
    hi  TermColor6          ctermfg=2
    hi  TermColor14         ctermfg=2

    " White and Light Gray
    hi  TermColor7          ctermfg=2
    hi  TermColor15         ctermfg=2
```

Thoughts on what I can do to help you debug this?

**Comment by justinmk** [comment]
Make sure you didn't enable termguicolors, else ctermfg would be ignored.

**Comment by jfelchner** [comment]
@justinmk but doesn't disabling `termguicolors` mean that I can't use true color themes?

**Comment by jfelchner** [comment]
Setting the colors like this:

```
    hi  TermColor0          guifg=#1b2b34
    hi  TermColor8          guifg=#65737e
```

Seems to be getting me slightly closer.  White works, but everything else is the "black" color.

**Comment by jfelchner** [comment]
Alright, so `TermColor0` is working, but none of the other ones seem to be working.  I applied:

```
let g:terminal_color_1  = '#b7eeb1'
highlight  TermColor1     guifg=#b7eeb1 ctermfg=DarkGreen
highlight  TermColor01    guifg=#b7eeb1 ctermfg=DarkGreen
```

For every number from `1` to `15` which should have made them all green, but no dice.

@justinmk it would be super helpful if I could use my syntax shortcut to see what highlight groups are being applied in the terminal buffer, but this doesn't seem to work:

`:echo "hi<" . synIDattr(synID(line("."),col("."),1),"name") . '> trans<' . synIDattr(synID(line("."),col("."),0),"name") . "> lo<" . synIDattr(synIDtrans(synID(line("."),col("."),1)),"name") . ">" . " FG:" . synIDattr(synIDtrans(synID(line("."),col("."),1)),"fg#")<CR>`

It returns empty results for any line in a term buffer.

**Comment by jfelchner** [comment]
@justinmk for what it's worth, I finally did get the `g:terminal_color_` versions to work for me.  If I used them alone, all works as expected.  However, if I apply any syntax groups for `TermColor`, everything turns to the `TermColor0` color and doesn't work properly.

**Comment by justinmk** [comment]
>  it would be super helpful if I could use my syntax shortcut to see what highlight groups are being applied in the terminal buffer

Yeah, that's not implemented in #7406. It would be nice but I'm sure not it's worth it: it's more work, could have a performance impact, and other Vim highlight groups like diff-mode/folds/etc. don't work like that either. 

>  I finally did get the g:terminal_color_ versions to work for me

Ok, but this issue was about changing the colors at runtime (after startup).

**Comment by jfelchner** [comment]
> Ok, but this issue was about changing the colors at runtime (after startup).

Absolutely. I'm just letting you know that it's not working as expected, whereas `g:terminal_color_` is.  Presumably I should be able to replace `g:terminal_color_` with `TermColor` but as of now, I cannot do that.

**Comment by justinmk** [comment]
> `g:terminal_color_*` is read only at terminal startup. ... attempting to change these variables doesn't cause any color change

This could be fixed by setting a dict watcher (`dictwatcheradd()`) in `terminal.c` which [reinitializes the color palette](https://github.com/neovim/neovim/blob/7bb858c39cac9b9af321fc90470c91d2e31ed96e/src/nvim/terminal.c#L225-L248). But it would be better to finish https://github.com/neovim/neovim/pull/7406 instead.

**Comment by bfredl** [comment]
#7406 would also enable output-buffer plugins to reuse "palette forwarding", i e when `TermColorFg4` is not overridden, then the plugin (rendering some ansi seq data) can use it to access color 4 in the terminal even with `termguicolors` set.

**Comment by xer0x** [comment]
This isn't the right place, but is there a way to make the colors in :terminal use colors from the current theme? It's picking the usual system-default style colors, and they are difficult to read.

**Comment by bfredl** [comment]
@xer0x This is planned #10994
